DllImportGenerator/DllImportGenerator/Analyzers/GeneratedDllImportAnalyzer.cs

```diff
@@ -64,8 +74,7 @@ private static void AnalyzeSymbol(SymbolAnalysisContext context, INamedTypeSymbo
                 foreach (var reference in methodSymbol.DeclaringSyntaxReferences)
                 {
                     var syntax = reference.GetSyntax(context.CancellationToken);
-                    var methodSyntax = syntax as MethodDeclarationSyntax;
-                    if (methodSyntax == null)
+                    if (syntax is not MethodDeclarationSyntax methodSyntax)
```

Comment:
I want to hear @elinor-fung's thoughts here but having a false statement have side effects really is odd and against almost any programming language model I'm familiar with. If we can reach consensus about this I will stop commenting on it but it is a difficult invariant for me to mentally accept.

Reply:
Another option would be to do something like the following:

```csharp
if (syntax is MethodDeclarationSyntax methodSyntax && !methodSyntax.Modifiers.Any(SyntaxKind.PartialKeyword))
{
     break;
}
```

The only reason I use it so much is that I'm trying to keep with our model of "check and bail out early to reduce curly brace nesting" while also using the most concise syntax, which in this case is not-patterns.

Reply:
>The only reason I use it so much is that I'm trying to keep with our model of "check and bail out early to reduce curly brace nesting"

Agree.

>while also using the most concise syntax, which in this case is not-patterns.

Unsure if we need the "most concise syntax". The `as` pattern has been around for 15 years now and I am dubious it is any more difficult to recall than paging in and accepting the number of side effects in each of these other approaches. I think this comes down to having a single extra line declaring a variable _or_ using a pattern that is anathema to programming logic, non-intuitive side-effects, or additional predicates using Linq that is much more code.

Reply:
For me paging back in the logic of `as` and `null` takes more effort than the natural language syntax of pattern matching.

For `as`, I have to think "This is a conditional cast that returns null, so the variable has to be nullable now. Ok now we're checking for null, so either the object is null or the cast failed. So we end up in the if block if the value is null or not that type."

For `not Type foo`, I think "Ok, we're checking if this object is a non-nullable Type. If it's not, we execute the code in the if. If it is, we execute the code in the else" which is a much more natural mental model for me. It's also the recommended model since it doesn't risk using an overloaded `==` or `!=` operator, which the common `as` pattern does. And it compiles down to the same thing as the `as` pattern using reference equality, so it's just as efficient.

In my opinion, the `as` pattern has more of a mental load and can end up having worse performance than the pattern matching case, but I know we strongly disagree when it comes to things like this.

Reply:
I have a lot of trouble with the pattern variables declared under `is not` as well. It is really the combined `not` + declaration that trips me up, so the alternate option @jkoritzinsky pointed out would be easier to understand for me - and my preference here.

It is combining the existing condition / nesting below, so I don't think it goes against our 'check and bail out early to reduce curly brace nesting' in this case.

Reply:
Sounds good.